firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,432 @@ use crate::{
-    merkle: &'a Merkle<S, T>,
-    mut children: Iter,
-    parents: &mut Vec<(NodeObjRef<'a>, u8)>,
-    node: &mut NodeRef<'a>,
-    pos: &mut u8,
-) -> Result<MustUse<bool>, super::MerkleError>
+/// Takes in an iterator over a node's partial path and an iterator over the
+/// unmatched portion of a key.
+/// The first returned element is:
+/// * [Ordering::Less] if the node is before the key.
+/// * [Ordering::Equal] if the node is a prefix of the key.
+/// * [Ordering::Greater] if the node is after the key.
+/// The second returned element is the unmatched portion of the key after the
+/// partial path has been matched.
+fn compare_partial_path<'a, I1, I2>(
+    partial_path_iter: I1,
+    mut unmatched_key_nibbles_iter: I2,
+) -> (Ordering, I2)
 where
-    Iter: Iterator<Item = (DiskAddress, u8)>,
-    S: ShaleStore<Node>,
+    I1: Iterator<Item = &'a u8>,
+    I2: Iterator<Item = u8>,
 {
-    if let Some((child_addr, child_pos)) = children.next() {
-        let child = merkle.get_node(child_addr)?;
-
-        *pos = child_pos;
-        let node = std::mem::replace(node, NodeRef::New(child));
-        parents.push((node.into_node(), *pos));
-    } else {
-        let Some((next_parent, next_pos)) = parents.pop() else {
-            return Ok(false.into());
+    for next_partial_path_nibble in partial_path_iter {
+        let Some(next_key_nibble) = unmatched_key_nibbles_iter.next() else {
+            return (Ordering::Greater, unmatched_key_nibbles_iter);
         };
 
-        *node = NodeRef::Visited(next_parent);
-        *pos = next_pos;
+        match next_partial_path_nibble.cmp(&next_key_nibble) {
+            Ordering::Less => return (Ordering::Less, unmatched_key_nibbles_iter),
+            Ordering::Greater => return (Ordering::Greater, unmatched_key_nibbles_iter),
+            Ordering::Equal => {}
+        }
     }
 
-    Ok(true.into())
+    (Ordering::Equal, unmatched_key_nibbles_iter)
 }
 
-/// create an iterator over the key-nibbles from all parents _excluding_ the sentinal node.
-fn nibble_iter_from_parents<'a>(parents: &'a [(NodeObjRef, u8)]) -> impl Iterator<Item = u8> + 'a {
-    parents
-        .iter()
-        .skip(1) // always skip the sentinal node
-        .flat_map(|(parent, child_nibble)| match parent.inner() {
-            NodeType::Branch(branch) => Either::Left(
-                branch
-                    .path
-                    .iter()
-                    .copied()
-                    .chain(std::iter::once(*child_nibble)),
-            ),
-            NodeType::Extension(extension) => Either::Right(extension.path.iter().copied()),
-            NodeType::Leaf(leaf) => Either::Right(leaf.path.iter().copied()),
-        })
+/// Returns an iterator that returns (child_addr, pos) for each non-empty child of `branch`,
+/// where `pos` is the position of the child in `branch`'s children array.
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
```

Comment:
nit: this seems a little odd, in that we had to reverse the output of `enumerate()`. Seems like these tuples should be `(u8, DiskAddress)` to match the way `enumerate()` works.

nit: naming could be better. Typically `get` is not used in rust. Some ideas:
- `enumerated_children_iter`
- `as_enumerated_children_iter`